Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -939,26 +970,388 @@ Query Serializer::DecodeQueryTarget(
-  StructuredQuery query = DecodeStructuredQuery(proto.structured_query);
+  const google_firestore_v1_StructuredQuery& query = proto.structured_query;
 
-  size_t from_count = query.from.size();
+  CollectionGroupId collection_group;
+  size_t from_count = query.from_count;
   if (from_count > 0) {
     if (from_count != 1) {
       reader->Fail(
           "StructuredQuery.from with more than one collection is not "
           "supported.");
     }
 
-    path = path.Append(query.from[0].collection_id);
+    google_firestore_v1_StructuredQuery_CollectionSelector& from =
+        query.from[0];
+    auto collection_id = DecodeString(from.collection_id);
+    if (from.all_descendants) {
+      collection_group = std::make_shared<const std::string>(collection_id);
+    } else {
+      path = path.Append(collection_id);
+    }
+  }
+
+  FilterList filter_by;
+  if (query.where.which_filter_type != 0) {  // FIXME
+    filter_by = DecodeFilters(reader, query.where);
+  }
+
+  OrderByList order_by;
+  if (query.order_by_count > 0) {
+    order_by = DecodeOrderBys(reader, query.order_by, query.order_by_count);
+  }
+
+  int32_t limit = Query::kNoLimit;
+  if (query.limit.value != Query::kNoLimit) {  // FIXME
+    limit = query.limit.value;
+  }
+
+  std::shared_ptr<Bound> start_at;
+  if (query.start_at.values_count > 0) {
+    start_at = DecodeBound(reader, query.start_at);
+  }
+
+  std::shared_ptr<Bound> end_at;
+  if (query.end_at.values_count > 0) {
+    end_at = DecodeBound(reader, query.end_at);
+  }
+
+  return Query(std::move(path), std::move(collection_group),
+               std::move(filter_by), std::move(order_by), limit,
+               std::move(start_at), std::move(end_at));
+}
+
+google_firestore_v1_StructuredQuery_Filter Serializer::EncodeFilters(
+    const FilterList& filters) const {
+  google_firestore_v1_StructuredQuery_Filter result{};
+  if (filters.empty()) {
+    // FIXME
+    return result;
+  }
+
+  std::vector<google_firestore_v1_StructuredQuery_Filter> converted_filters;
+  for (const auto& filter : filters) {
+    if (filter.IsAFieldFilter()) {
+      converted_filters.push_back(
+          EncodeNonCompositeFilter(FieldFilter{filter}));
+    }
+  }
+  if (converted_filters.size() == 1) {
+    // Special case: no existing filters and we only need to add one filter.
+    // This can be made the single root filter without a composite filter.
+    return converted_filters.back();
+  }
+
+  result.which_filter_type =
+      google_firestore_v1_StructuredQuery_Filter_composite_filter_tag;
+  google_firestore_v1_StructuredQuery_CompositeFilter& composite =
+      result.composite_filter;
+  composite.op =
+      google_firestore_v1_StructuredQuery_CompositeFilter_Operator_AND;
+
+  auto count = CheckedSize(converted_filters.size());
+  composite.filters_count = count;
+  composite.filters =
+      MakeArray<google_firestore_v1_StructuredQuery_Filter>(count);
+  for (pb_size_t i = 0; i != count; ++i) {
```

Comment:
I did it mostly for consistency with Objective-C.

Agreed that it's inefficient, though there's no getting around of having to traverse the list of filters once beforehand to know how many field filters it contains (so that it is known how much size to allocate for the array). Took a stab at doing this in a more efficient manner, PTAL.